File: ios/Classes/Core/GameEngineRegistry.swift
import Flutter
import Foundation

/**
 * Protocol for engine factories
 *
 * Engine plugins must provide a factory that creates their specific controller.
 */
public protocol GameEngineFactory {
    /// Create an engine controller
    func createController(
        frame: CGRect,
        viewId: Int64,
        messenger: FlutterBinaryMessenger,
        config: [String: Any]
    ) -> GameEnginePlatformView
}

/**
 * Singleton registry for game engine implementations
 *
 * Manages the registration and lifecycle of engine controllers and factories.
 */
public class GameEngineRegistry {

    // MARK: - Singleton

    public static let shared = GameEngineRegistry()

    private init() {}

    // MARK: - Properties

    private var factories: [String: GameEngineFactory] = [:]
    private var controllers: [GameEnginePlatformView] = []

    // MARK: - Factory Management

    /// Register an engine factory
    public func registerFactory(engineType: String, factory: GameEngineFactory) {
        factories[engineType] = factory
    }

    /// Unregister an engine factory
    public func unregisterFactory(engineType: String) {
        factories.removeValue(forKey: engineType)
    }

    /// Check if an engine is registered
    public func isEngineRegistered(_ engineType: String) -> Bool {
        return factories[engineType] != nil
    }

    /// Get all registered engine types
    public func getRegisteredEngines() -> [String] {
        return Array(factories.keys)
    }

    /// Get factory for a specific engine type
    public func getFactory(_ engineType: String) -> GameEngineFactory? {
        return factories[engineType]
    }

    // MARK: - Controller Management

    /// Register a controller instance
    public func registerController(_ controller: GameEnginePlatformView) {
        controllers.append(controller)
    }

    /// Unregister a controller instance
    public func unregisterController(_ controller: GameEnginePlatformView) {
        controllers.removeAll { $0 === controller as AnyObject }
    }

    /// Get all active controllers
    public func getControllers() -> [GameEnginePlatformView] {
        return controllers
    }

    // MARK: - Lifecycle Management

    /// Called when app becomes active
    public func onAppBecameActive() {
        // Notify controllers if needed
    }

    /// Called when app will resign active
    public func onAppWillResignActive() {
        // Notify controllers if needed
    }

    // MARK: - Cleanup

    /// Clear all registrations (mainly for testing)
    public func clear() {
        factories.removeAll()
        controllers.removeAll()
    }
}

/**
 * Platform view factory for game engines
 *
 * Wraps the GameEngineFactory protocol for Flutter's platform view system.
 */
public class GameEnginePlatformViewFactory: NSObject, FlutterPlatformViewFactory {

    private let messenger: FlutterBinaryMessenger
    private let engineType: String

    public init(messenger: FlutterBinaryMessenger, engineType: String) {
        self.messenger = messenger
        self.engineType = engineType
        super.init()
    }

    public func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        let config = args as? [String: Any] ?? [:]

        guard let factory = GameEngineRegistry.shared.getFactory(engineType) else {
            fatalError("Engine \(engineType) is not registered")
        }

        let controller = factory.createController(
            frame: frame,
            viewId: viewId,
            messenger: messenger,
            config: config
        )

        GameEngineRegistry.shared.registerController(controller)
        return controller
    }

    public func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

File: android/src/main/kotlin/com/xraph/gameframework/gameframework/core/GameEngineRegistry.kt
package com.xraph.gameframework.gameframework.core

import android.app.Activity
import androidx.lifecycle.Lifecycle

/**
 * Singleton registry for game engine implementations
 *
 * Manages the registration and lifecycle of engine controllers and factories.
 * Ensures only one engine can be active at a time.
 */
class GameEngineRegistry private constructor() {

    private val factories = mutableMapOf<String, GameEngineFactory>()
    private val controllers = mutableListOf<GameEngineController>()
    private var currentActivity: Activity? = null
    private var currentLifecycle: Lifecycle? = null

    companion object {
        @Volatile
        private var INSTANCE: GameEngineRegistry? = null

        val instance: GameEngineRegistry
            get() = INSTANCE ?: synchronized(this) {
                INSTANCE ?: GameEngineRegistry().also { INSTANCE = it }
            }
    }

    /**
     * Register an engine factory for a specific engine type
     *
     * Called by engine plugins during initialization.
     *
     * @param engineType The engine identifier (e.g., "unity", "unreal")
     * @param factory The factory that creates controllers for this engine
     */
    fun registerFactory(engineType: String, factory: GameEngineFactory) {
        factories[engineType] = factory
    }

    /**
     * Unregister an engine factory
     */
    fun unregisterFactory(engineType: String) {
        factories.remove(engineType)
    }

    /**
     * Check if an engine is registered
     */
    fun isEngineRegistered(engineType: String): Boolean {
        return factories.containsKey(engineType)
    }

    /**
     * Get all registered engine types
     */
    fun getRegisteredEngines(): List<String> {
        return factories.keys.toList()
    }

    /**
     * Get factory for a specific engine type
     */
    fun getFactory(engineType: String): GameEngineFactory? {
        return factories[engineType]
    }

    /**
     * Register a controller instance
     *
     * Called automatically by factories when creating controllers.
     */
    fun registerController(controller: GameEngineController) {
        controllers.add(controller)
    }

    /**
     * Unregister a controller instance
     */
    fun unregisterController(controller: GameEngineController) {
        controllers.remove(controller)
    }

    /**
     * Get all active controllers
     */
    fun getControllers(): List<GameEngineController> {
        return controllers.toList()
    }

    /**
     * Notify all controllers of activity attachment
     */
    fun onActivityAttached(activity: Activity, lifecycle: Lifecycle) {
        currentActivity = activity
        currentLifecycle = lifecycle
    }

    /**
     * Notify all controllers of activity detachment
     */
    fun onActivityDetached() {
        currentActivity = null
        currentLifecycle = null
    }

    /**
     * Get the current activity
     */
    fun getCurrentActivity(): Activity? = currentActivity

    /**
     * Get the current lifecycle
     */
    fun getCurrentLifecycle(): Lifecycle? = currentLifecycle

    /**
     * Clear all registrations (mainly for testing)
     */
    fun clear() {
        factories.clear()
        controllers.clear()
        currentActivity = null
        currentLifecycle = null
    }
}
